rust/kernel/net/device.rs

```diff
@@ -0,0 +1,1076 @@
+    use core::ops::{Add, AddAssign, Deref, DerefMut, Sub, SubAssign};
+
+    /// Holds multiple flags to give to an interface via [`super::NetDevice::add_flag`].
+    #[repr(transparent)]
+    #[derive(Debug, Clone, Copy)]
+    pub struct NetIF(u64);
+
+    impl NetIF {
+        /// Create new Flag with value `0`.
+        pub const fn new() -> Self {
+            Self(0)
+        }
+
+        /// Add flag to Self.
+        pub fn add_flag(&mut self, flag: u64) {
+            self.0 |= flag;
+        }
+
+        /// Remove the given flag from Self.
+        pub fn remove_flag(&mut self, flag: u64) {
+            self.0 &= !(flag);
+        }
+    }
+
+    impl Deref for NetIF {
+        type Target = u64;
+
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+
+    impl DerefMut for NetIF {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut self.0
+        }
+    }
+
+    impl Add for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn add(self, rhs: Self) -> Self::Output {
+            Self::from(self.0 | rhs.0)
+        }
+    }
+
+    impl Add<u64> for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn add(self, rhs: u64) -> Self::Output {
+            Self::from(self.0 | rhs)
+        }
+    }
+
+    impl Sub for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn sub(self, rhs: Self) -> Self::Output {
+            Self::from(self.0 & !rhs.0)
+        }
+    }
+
+    impl Sub<u64> for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn sub(self, rhs: u64) -> Self::Output {
+            Self::from(self.0 & !rhs)
+        }
+    }
+
+    impl AddAssign for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn add_assign(&mut self, rhs: Self) {
+            self.0 |= rhs.0
+        }
+    }
+
+    impl AddAssign<u64> for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn add_assign(&mut self, rhs: u64) {
+            self.0 |= rhs
+        }
+    }
+
+    impl SubAssign for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn sub_assign(&mut self, rhs: Self) {
+            self.0 &= !rhs.0
+        }
+    }
+
+    impl SubAssign<u64> for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn sub_assign(&mut self, rhs: u64) {
+            self.0 &= !rhs
+        }
+    }
+
+    impl From<u64> for NetIF {
+        fn from(flags: u64) -> Self {
+            Self(flags)
+        }
+    }
+
+    impl From<NetIF> for u64 {
+        fn from(flag: NetIF) -> Self {
+            flag.0
+        }
+    }
+
+    macro_rules! _netif_f {
+        ($name:ident, $binding:ident) => {
+            /// [`NetIF`] flag for $binding
+            pub const $name: u64 = 1u64 << $crate::bindings::$binding;
```

Comment:
This should be associated constants, and have a type `NetIF` rather than `u64`.

Reply:
Tried that, rustc paniced, as 3 is not 4 *shrug*.

Reply:
Panicked as in gave an ICE (internal compiler error)? If so please open an issue on rust-lang/rust.

Reply:
Yes. Because of the bitshift. Which would need const functions. That panicked as an unstable feature.

Reply:
I tried locally and it works just fine.
```diff
diff --git a/rust/kernel/net/device.rs b/rust/kernel/net/device.rs
index c4a8a7dec285..62d80126731c 100644
--- a/rust/kernel/net/device.rs
+++ b/rust/kernel/net/device.rs
@@ -972,17 +972,18 @@ pub mod feature {
     macro_rules! _netif_f {
         ($name:ident, $binding:ident) => {
             /// [`NetIF`] flag for $binding
-            pub const $name: u64 = 1u64 << $crate::bindings::$binding;
+            pub const $name: NetIF = NetIF(1u64 << $crate::bindings::$binding);
         };
     }
 
     macro_rules! _netif_f_sum {
         ($name:ident, $($f:ident),+) => {
             /// [`NetIF`] flag for $name
-            pub const $name: u64 = $($crate::net::device::feature::$f |)* 0;
+            pub const $name: NetIF = NetIF($(NetIF::$f.0)|*);
         };
     }
 
+    impl NetIF {
     _netif_f!(NETIF_F_SG, NETIF_F_SG_BIT);
     _netif_f!(NETIF_F_FRAGLIST, NETIF_F_FRAGLIST_BIT);
     _netif_f!(NETIF_F_TSO, NETIF_F_TSO_BIT);
@@ -1025,6 +1026,7 @@ pub mod feature {
         NETIF_F_GSO_UDP_TUNNEL,
         NETIF_F_GSO_UDP_TUNNEL_CSUM
     );
+    }
 }
 
 /// Helper functions for NetDevices.
```

Reply:
Will apply and let's see what the pipeline says.

Reply:
I haven't adjusted the use sites in dummy_rs.

Reply:
Builds on my machine as well. I think I tried a different approach as I did this.

Reply:
I just noticed that `$binding` and `$name` does not get expanded in DocStrings. I don't have a clue on how to fix that. Also it's not nice, but I think we can live with it.

Reply:
I think you can use something like `#[doc = concat!("Foo", $binding, "bar")]`. That is what libcore does.

Reply:
~~~
error: unexpected token: `(/*ERROR*/)`
    --> ../rust/kernel/net/device.rs:981:21
     |
981  |               #[doc = concat!("[`NetIF`] flag for ", $name)]
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1020 | /         _netif_f_sum!(
1021 | |             NETIF_F_GSO_ENCAP_ALL,
1022 | |             NETIF_F_GSO_GRE,
1023 | |             NETIF_F_GSO_GRE_CSUM,
...    |
1027 | |             NETIF_F_GSO_UDP_TUNNEL_CSUM
1028 | |         );
     | |__________- in this macro invocation
     |
     = note: this error originates in the macro `_netif_f_sum` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 42 previous errors
~~~
It's not so easy :-)

Reply:
You need to use `core::stringify!($name)`

Reply:
`stringify!()` helped. Now it works. 

The output is in big letters (e.g. `NetIF flag for NETIF_F_TSO_MANGLEID_BIT`). But I don't think that is bad. Of course I would be nice do remove BIT, but I think it's not important, and It's readable as is.

Reply:
Github... @nbdd0121 it does not show your comment about stringify. only saw it on my phone, not in the web ui :-(